### PR TITLE
press key h to clear current node analysis.

### PR DESCRIPTION
**Pull request by xiaoyifang**
issue https://github.com/sanderland/katrain/issues/330
does not sure if there is a more better solution.haven't program in python for years.

### CODE REVIEW

**Comment by sanderland** [comment]
please target the next release, not master. 
I think this conflates two features and should be split up:

* reset analysis to base state (whatever key is next on the middle row)
* stop current analysis (esc?, keep in mind to also turn off continuous analysis)

**Comment by xiaoyifang** [comment]
does the feature "reset analysis to base state"  contains the feature "stop current analysis "?

**Comment by sanderland** [comment]
> does the feature "reset analysis to base state" contains the feature "stop current analysis "?

It is hard to determine what queries belong to the current analysis, so I wouldn't think so. You're right that there's some annoying edge cases though.

**Comment by xiaoyifang** [comment]
the branch contribute  can not start ,when run ,it will throw an exception 
"ERROR: Engine died unexpectedly without sending output, possibly due to out of memory: status 3221225501ERROR: Engine died unexpectedly without sending output, possibly due to out of memory: status 3221225501
"

**Comment by xiaoyifang** [comment]
maybe use new field in GameNode such as ,query_keys to store all the current analysis queryid.
i want to bind the clear function to a key ,any recommend keys .

**Comment by sanderland** [comment]
> maybe use new field in GameNode such as ,query_keys to store all the current analysis queryid.
maybe it'd be cleaner to store an optional extra field in the tuple in queries. then terminate queries can take an optional node and either terminate queries for that node, or all.

> i want to bind the clear function to a key ,any recommend keys .

h?

**Comment by sanderland** [comment]
just use the request review feature when you're ready @xiaoyifang

**Comment by xiaoyifang** [comment]
> just use the request review feature when you're ready @xiaoyifang

I'm ready,   but does not know how to  start request review.

**Comment by sanderland** [comment]
image

File: katrain/core/engine.py [hunk 1]
```diff
@@ -145,6 +146,13 @@ def on_new_game(self):
                 self.queries = {}
                 self.write_queue = queue.Queue()
 
+    def terminate_current_queries(self):
+        with self.thread_lock:
+            for query_id in list(self.queries.keys()):
+                callback, error_callback, start_time, next_move, cn = self.queries.pop(query_id, None)
```
Comment (sanderland) [inline katrain/core/engine.py:152]: cn -> node

File: katrain/core/game.py [hunk 2]
```diff
@@ -438,6 +438,13 @@ def undo(self, n_times=1, stop_on_mistake=None):
             return
         super().undo(n_times=n_times, stop_on_mistake=stop_on_mistake)
 
+    def reset(self):
+        cn = self.current_node
+        engine = self.engines[cn.next_player]
+        engine.terminate_current_queries()
```
Comment (sanderland) [inline katrain/core/game.py:444]: terminate_queries(cn)

File: katrain/core/engine.py [hunk 3]
```diff
@@ -311,7 +320,9 @@ def request_analysis(
         next_move: Optional[GameNode] = None,
         extra_settings: Optional[Dict] = None,
         report_every: Optional[float] = None,
+        current_analysis: bool = False,
     ):
+        self.analysis_node = analysis_node
```
Comment (sanderland) [inline katrain/core/engine.py:325]: remove, the engine may be working on any number of nodes and the last query is not necessarily the currnet node

File: katrain/core/engine.py [hunk 4]
```diff
@@ -378,5 +389,5 @@ def request_analysis(
             query["reportDuringSearchEvery"] = report_every
         if avoid:
             query["avoidMoves"] = avoid
-        self.send_query(query, callback, error_callback, next_move)
+        self.send_query(query, callback, error_callback, next_move, current_analysis)
```
Comment (sanderland) [inline katrain/core/engine.py:392]: pass analysis node instead

**Review by sanderland — CHANGES_REQUESTED** [review]
Have another look at the logic, keeping in mind that the engine is highly parallel

File: katrain/core/engine.py [hunk 5]
```diff
@@ -146,11 +146,11 @@ def on_new_game(self):
...
                 self.write_queue = queue.Queue()
 
-    def terminate_current_queries(self):
+    def terminate_queries(self, current_node=None):
         with self.thread_lock:
             for query_id in list(self.queries.keys()):
-                callback, error_callback, start_time, next_move, cn = self.queries.pop(query_id, None)
-                if cn is not None and self.analysis_node == cn:
+                callback, error_callback, start_time, next_move, node = self.queries.pop(query_id, None)
+                if current_node is not None and current_node == node:
```
Comment (sanderland) [inline katrain/core/engine.py:152]: current_node is None or

File: katrain/core/engine.py [hunk 6]
```diff
@@ -281,25 +281,24 @@ def _analysis_read_thread(self):
     def _write_stdin_thread(self):  # flush only in a thread since it returns only when the other program reads
         while self.katago_process is not None:
             try:
-                query, callback, error_callback, next_move, current_analysis = self.write_queue.get(block=True, timeout=0.1)
+                query, callback, error_callback, next_move, current_node = self.write_queue.get(block=True, timeout=0.1)
```
Comment (sanderland) [inline katrain/core/engine.py:283]: just `node`

File: katrain/core/engine.py [hunk 7]
```diff
@@ -281,25 +281,24 @@ def _analysis_read_thread(self):
...
                 self.katrain.log(f"Sending query {query['id']}: {json.dumps(query)}", OUTPUT_DEBUG)
                 try:
                     self.katago_process.stdin.write((json.dumps(query) + "\n").encode())
                     self.katago_process.stdin.flush()
                 except OSError as e:
                     self.check_alive(os_error=str(e), exception_if_dead=False)
 
-    def send_query(self, query, callback, error_callback, next_move=None, current_analysis=False):
-        self.write_queue.put((query, callback, error_callback, next_move, current_analysis))
+    def send_query(self, query, callback, error_callback, next_move=None, current_node=None):
```
Comment (sanderland) [inline katrain/core/engine.py:299]: `node`

**Review by sanderland — CHANGES_REQUESTED** [review]
one tiny thing, will merge later when I have the time to test.

**Comment by sanderland** [comment]
Thank you for your contribution @xiaoyifang

**Comment by sanderland** [comment]
gui + some renaming in 5011d07

**Pull request merged**